mypyc/irbuild/function.py

```diff
@@ -857,3 +858,24 @@ def gen_dispatch_func_ir(
     func_decl = FuncDecl(dispatch_name, None, builder.module_name, sig)
     dispatch_func_ir = FuncIR(func_decl, args, blocks)
     return dispatch_func_ir
+
+
+def sort_with_subclasses_first(
+    impls: List[Tuple[TypeInfo, FuncDef]]
+) -> Iterable[Tuple[TypeInfo, FuncDef]]:
+    def is_subclass(typ1: TypeInfo, typ2: TypeInfo) -> bool:
+        return typ2 in typ1.mro
+    dispatch_types: List[TypeInfo] = []
+    funcs: List[FuncDef] = []
+    for typ, impl in impls:
+        # If this type is a subclass of anything we've seen previously, put it in the front so it
+        # gets checked first
+        if any(is_subclass(typ, prev_type) for prev_type in dispatch_types):
+            dispatch_types.insert(0, typ)
+            funcs.insert(0, impl)
```

Comment:
I think this logic is flawed. Imagine a situation like this:
```python
class A: ...
class B(A): ...
class C(B): ...
```
Then if these classes appear in `impls` in the order `A, C, B`, then after this "sorting", they will be in the order `B, C, A`, which is wrong.

Also when fixing this I think it could make sense to make it stable (i.e. order of unrelated items should be preserved).

Reply:
I think this should be fixed now. Could you take a look and let me know what you think?